### ISSUE TITLE
Clones

### ISSUE THREAD

**Issue by casey** [soenkehahn/cradle#52]
I think that there are additional clones in cradle (or will be once #46 lands) on top of the clones done when using `std::process::Command`.

I looked at the source of `std::process::Command`, and on Unix it will create `CString`s for all arguments. (`CString` is an owned, null-terminated string, suitable for passing to system calls.)

`cradle` will do additional allocations on top of that, so it would be nice to either avoid that. It might not be possible, or it might complicate the API, for example by adding lifetimes to a bunch of things.

**Comment by soenkehahn** [comment]
Overall I'm not too worried about cloning executable names and command line arguments. Even in pretty extreme cases, I would think that these should only be a couple of kilobytes. (IIUC my OS also imposes a limit on bytes in command line arguments of roughly 2MB.) I don't think I would want to avoid cloning if that comes with any negative impact on `cradle`s api.

Of course, if there's no impact it'd be great to avoid cloning.

**Comment by casey** [comment]
Would changing the `CmdArgument` trait to take `self` by value work:

```rust
pub trait CmdArgument {
    #[doc(hidden)]
    fn prepare_config(&self, config: &mut Config);
}
```

Then, there's no need to provide the implementation for `&T` that requires cloning.

I was thinking that maybe the further clone when the item is pushed into the arguments couldn't be avoided, due to temporaries, but now I'm not so sure.

This works:

```rust
#[derive(Default)]
struct Command<'a> {
    args: Vec<&'a str>,
}

impl<'a> Command<'a> {
    fn arg(mut self, arg: &'a str) -> Self {
        self.args.push(arg);
        self
    }

    fn run(self) {
        std::process::Command::new(self.args[0])
            .args(&self.args[1..])
            .status()
            .unwrap();
    }
}

fn main() {
    Command::default().arg("echo").arg(&format!("foo")).run();
}
```

The argument to `echo` is a temporary `String`, but because the temporary is created and used in the same expression, the compiler knows not to drop it until it's finished evaluating the expression.

For `cradle`, I think the same thing applies. The macros are all going to boil down to a single expression which is then immediately consumed, so `Config` could, I think, look like this:

```rust
pub struct Config<'a> {
    pub(crate) arguments: Vec<&'a OsStr>,
    …
}
```

**Comment by casey** [comment]
I gave this another shot, and found it extremely fiddly to implement.

The main issue I ran into is that in order to avoid lifetime errors, all macros need to expand into a single statement, so that temporaries passed to `cmd!(…)` aren't deallocated early.

So A and Everything Else in this:

```rust
    (config: $config:ident, args: $head:expr, $($tail:tt)*) => {
        $crate::CmdArgument::prepare_config($head, &mut $config); // Stmt. A
        $crate::prepare_config!(config: $config, args: $($tail)*); // Stmt. Everything Else
    };
```

Would have to be refactored to be a single weird statement containing both A and Everything Else. I think maybe it could be done, but it probably isn't worth it just for an optimization.
